Adding libraries for connected components and bipartiteness

Zainab-Abbas opened the pull request:
Pull request for library package and tests of connected components and bipatiteness check

Zainab-Abbas committed:
Adding libraries for connected components and bipartiteness

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.graph.streaming.example.library;
```
We should have a separate package for examples and library methods.

Zainab-Abbas replied:
you mean like this right ?
package org.apache.flink.graph.streaming.library

senorcarbone replied:
yes @Zainab-Abbas that is fine (like in Gelly). It's a general package that the examples are also using so it should be under graph.streaming

vasia replied:
yes!

Zainab-Abbas replied:
alright thanks @senorcarbone @vasia will make it separate.

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+import java.io.Serializable;
+
+/**
+ * The bipartiteness check example tests whether an input graph is bipartite
```
This one would be the library method, not an example

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+ * groups, such as no two nodes inside the same group is connected by an edge.
+ * The example uses the merge-tree abstraction of our graph streaming API.
+ */
+public class BipartitenessCheck <K extends Serializable,EV> extends WindowGraphAggregation<K,EV,Candidates, Candidates> implements Serializable {
```
Make sure there always exists a space after comma.

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+public class BipartitenessCheck <K extends Serializable,EV> extends WindowGraphAggregation<K,EV,Candidates, Candidates> implements Serializable {
+
+    public BipartitenessCheck() {
+        super( new updFun(), new combineFun (), new Candidates(true), 500, false);
```
Fix formatting

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+ */
+public class BipartitenessCheck <K extends Serializable,EV> extends WindowGraphAggregation<K,EV,Candidates, Candidates> implements Serializable {
+
+    public BipartitenessCheck() {
```
What do the parameters passed to the constructor represent? Better make these configurable and add a javadoc to describe what they do.

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,70 @@
+
+    @SuppressWarnings("serial")
+
+    public static class updFun <K extends Serializable> implements EdgesFold<Long, NullValue, Candidates> {
```
updFun => updateFunction
Could you also add a description?

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+}
+
```
Remove empty lines

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,78 @@
+    private long mergeWindowTime;
+
+
+    public ConnectedComponents(long mergeWindowTime) {
```
Add javadoc

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/BipartitenessCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+        GraphStream<Long, NullValue, NullValue> graph = new SimpleEdgeStream<>(edges, env);
+        DataStream<Candidates> cc = graph.aggregate(new BipartitenessCheck<Long,NullValue>());
+        cc.writeAsText(resultPath);
+        cc.print();
```
Why do you both print and write to a file?

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/BipartitenessCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    }
+
+    public static final List<Edge<Long, NullValue>> getEdges() {
```
Can you also add a test where the graph is not bipartite?

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/ConnectedComponentsTest.java:
```diff
@@ -0,0 +1,99 @@
+        GraphStream<Long, NullValue, NullValue> graph = new SimpleEdgeStream<>(edges, env);
+        DataStream<DisjointSet<Long>> cc = graph.aggregate(new ConnectedComponents<Long, NullValue>(5));
+        cc.writeAsText(resultPath);
+        cc.print();
```
remove printing?

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/ConnectedComponentsTest.java:
```diff
@@ -0,0 +1,99 @@
+        return edges;
+    }
+
+    public static final String[] Parser (ArrayList<String>list){
```
=> parser

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/ConnectedComponentsTest.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public static final String[] Parser (ArrayList<String>list){
+        int s= list.size();
```
formatting

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/ConnectedComponentsTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    public static final String[] Parser (ArrayList<String>list){
+        int s= list.size();
+        String R = list.get(s-1);  // to get the final combine result which is stored at the end of result
```
R => r

vasia commented on src/test/java/org/apache/flink/graph/streaming/example/test/ConnectedComponentsTest.java:
```diff
@@ -0,0 +1,99 @@
+        String R = list.get(s-1);  // to get the final combine result which is stored at the end of result
+        String t = null;
+        list.clear();
+        String [] G = R.split("=");
```
formatting

vasia commented:
Thanks for the PR @Zainab-Abbas! I've left some comments for improvements. Let us know if you need help.

Zainab-Abbas commented:
Thank you @vasia  for the comments i will fix them soon, i will ask incase i got stuck

senorcarbone commented:
:+1:  @Zainab-Abbas good job! Please cleanup the original examples as well. As far as I can see there is yet another ConnectedComponents.java under examples which is unchanged. You can rename that into `ConnectedComponentsExample`  or something and refactor it to use the actual library :)

Zainab-Abbas commented:
@senorcarbone thanks :+1:  will do that soon :)

Zainab-Abbas committed:
made formatting changes

Zainab-Abbas commented:
- Made changes to the connected components and bipartiteness examples
- Added library functions for connected components and bipartiteness check
- Added tests for connected components, bipartiteness and non-bipartiteness check
  -Added javadocs for different fuctions

vasia commented on src/main/java/org/apache/flink/graph/streaming/example/ConnectedComponentsExample.java:
```diff
@@ -147,52 +118,52 @@ private static boolean parseParameters(String[] args) {
 	public static final class IdentityFold implements FoldFunction<Tuple2<Long, Long>, Tuple2<Long, Long>> {
 		public Tuple2<Long, Long> fold(Tuple2<Long, Long> accumulator,
-				Tuple2<Long, Long> value) throws Exception {
+									   Tuple2<Long, Long> value) throws Exception {
```
indention seems off here

Zainab-Abbas replied:
changed indention for this one

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+import java.io.Serializable;
+
+/**
+ * The bipartiteness check library to check whether an input graph is bipartite
```
Rephrase this sentence to something like "The bipartiteness check library method checks whether ..."

Zainab-Abbas replied:
👍  done this

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+/**
+ * The bipartiteness check library to check whether an input graph is bipartite
+ * or not. A bipartite graph's vertices can be separated into two disjoint
+ * groups, such as no two nodes inside the same group is connected by an edge.
```
no two nodes... _are_ connected...

Zainab-Abbas replied:
done this 👍

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+ * The bipartiteness check library to check whether an input graph is bipartite
+ * or not. A bipartite graph's vertices can be separated into two disjoint
+ * groups, such as no two nodes inside the same group is connected by an edge.
+ * The library uses the merge-tree abstraction of our graph streaming API.
```
I don't think it does :)
It looks like window graph aggregation to me.

Zainab-Abbas replied:
changed it to:
The library uses the Window Graph Aggregation class of our graph streaming API.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+		 * @param edgeVal    the edge value
+		 * @return The data stream that is the result of applying the foldEdges function to the graph window.
+		 * @throws Exception
+		 */
```
Is this the javadoc of the `foldEdges` method?

Zainab-Abbas replied:
yes i used the javadoc of foldEdges method previously 
now changed it to :
Implements foldEdges method of EdgesFold interface for combining
two edges values into same type using merge method of the Candidates class.
In this case it updates the Bipartiteness value in each partition.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+	 * <p>
+	 * <p>
+	 * Like all functions, the ReduceFunction needs to be serializable, as defined in {@link java.io.Serializable}.
+	 */
```
Add a javadoc that describes the implementation, not the ReduceFunction interface.

Zainab-Abbas replied:
changed to :
Implements the ReduceFunction Interface, applies reduce function to
combine group of elements into a single value.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -0,0 +1,124 @@
+		 * @return The combined value of both input values.
+		 * @throws Exception This method may throw exceptions. Throwing an exception will cause the operation
+		 *                   to fail and may trigger recovery.
+		 */
```
Same here.

Zainab-Abbas replied:
changed to :
Implements reduce method of ReduceFunction interface.
Two values of Candidates class are combined into one using merge method
of the Candidate class.
In this case the merge method takes Bipartiteness values from different
partitions and merges them into one.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,129 @@
+
+	private long mergeWindowTime;
+
+	public ConnectedComponents(long mergeWindowTime) {
```
add javadoc and explain the `mergeWindowTime` argument

Zainab-Abbas replied:
changed to :
     Creates a ConnectedComponents object using WindowGraphAggregation class.
     This helps dividing the edge stream into parallel window for each partition.
     @param mergeWindowTime Window time in millisec for the merger.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,129 @@
+		 * @param edgeValue the edge value
+		 * @return The data stream that is the result of applying the foldEdges function to the graph window.
+		 * @throws Exception
+		 */
```
Same as the previous comments. Copying the javadoc isn't helpful. You should describe what this implementation is doing instead :)

Zainab-Abbas replied:
Changed to:
Implements foldEdges method of EdgesFold interface for combining
 two edges values into same type using union method of the DisjointSet class.
 In this case it updates the Connected Component value in each partition.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,129 @@
+	 * <p>
+	 * Like all functions, the ReduceFunction needs to be serializable, as defined in {@link java.io.Serializable}.
+	 */
+	public static class CombineCC<K extends Serializable> implements ReduceFunction<DisjointSet<K>> {
```
Same here...

Zainab-Abbas replied:
changed to:
Implements the ReduceFunction Interface, applies reduce function to
combine group of elements into a single value.

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/ConnectedComponents.java:
```diff
@@ -0,0 +1,129 @@
+		 *                   to fail and may trigger recovery.
+		 */
+		@Override
+		public DisjointSet<K> reduce(DisjointSet<K> s1, DisjointSet<K> s2) throws Exception {
```
and here

Zainab-Abbas replied:
changed to:
Implements reduce method of ReduceFunction interface.
Two values of DisjointSet class are combined into one using merge method
of the DisjointSet class.
 In this case the merge method takes Connected Components values from different
partitions and merges them into one.

vasia commented:
Thanks for updating the PR @Zainab-Abbas! I left a few comments. Let me know if you need help.

Zainab-Abbas commented:
Thank you for the comments @vasia 
I have replied to you comments with the changes i made :)
let me know if i am wrong some where

vasia commented:
Hey @Zainab-Abbas,
I see your comments but I don't see your changes... Did you forget to push your commit? :)

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

Zainab-Abbas commented:
@vasia just pushed my commit now :)

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -41,6 +41,12 @@
 
+	/**
+	 * Creates a BipartitenessCheck object using WindowGraphAggregation class.
+	 * This helps dividing the edge stream into parallel window for each partition.
```
"This helps dividing..." What does _this_ refer to?
also, "parallel windows*"

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -57,17 +63,17 @@ public static Candidates edgeToCandidate(long v1, long v2) throws Exception {
-	 * Implements EdgesFold Interface, applies a function to a vertex neighborhood
-	 * in the {@link GraphWindowStream#foldNeighbors(Object, EdgesFold)} method.
+	 * Implements the EdgesFold Interface, applies foldEdges function to
+	 * a vertex neighborhood
```
This is too general and not very informative. What does updateFunction compute in this algorithm specifically?

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -57,17 +63,17 @@ public static Candidates edgeToCandidate(long v1, long v2) throws Exception {
-		 * until only a single value remains.
+		 * Implements foldEdges method of EdgesFold interface for combining
+		 * two edges values into same type using merge method of the Candidates class.
+		 * In this case it updates the Bipartiteness value in each partition.
```
What do you mean by "Bipartiteness value"? Can you be a bit more specific?

vasia commented on src/main/java/org/apache/flink/graph/streaming/library/BipartitenessCheck.java:
```diff
@@ -83,31 +89,17 @@ public Candidates foldEdges(Candidates candidates, Long v1, Long v2, NullValue e
-	 * <p>
-	 * Like all functions, the ReduceFunction needs to be serializable, as defined in {@link java.io.Serializable}.
+	 * Implements the ReduceFunction Interface, applies reduce function to
+	 * combine group of elements into a single value.
```
That's again quite general. What does this method compute in this particular algorithm?

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

Zainab-Abbas committed:
Adding tests for connected components and bipartiteness

Zainab-Abbas commented:
@vasia Made some changes in the javadocs to explain more about the functions :) 
let me know if it is correct

vasia commented:
Thank you @Zainab-Abbas. I will merge this :)

Zainab-Abbas commented:
@vasia good 👍  you are welcome :)
